content/articles/how-to-build-a-web-crawler-with-java/index.md

```diff
@@ -0,0 +1,178 @@
+                String inputLine = in.readLine();
+
+                // read every line of the HTML content in the URL
+                // and concatenate each line to the rawHTML string until every line is read.
+                while(inputLine  != null){
+                    rawHTML += inputLine;
+
+                    inputLine = in.readLine();
+                }
+                in.close();
+            }catch (Exception e){
+                e.printStackTrace();
+            }
+
+
+            // create a regex pattern matching a URL
+            // that will validate the content of HTML in search of a URL.
+            String urlPattern = "(www|http:|https:)+[^\s]+[\\w]";
+            Pattern pattern = Pattern.compile(urlPattern);
+            Matcher matcher = pattern.matcher(rawHTML);
+
+            // Each time the regex matches a URL in the HTML,
+            // add it to the queue for the next traverse and to the list of visited URLs.
+            while(matcher.find()){
+                String actualURL = matcher.group();
+
+                if(!visitedURLs.contains(actualURL)){
+                    visitedURLs.add(actualURL);
+                    System.out.println("Website found with URL " + actualURL);
+                    urlQueue.add(actualURL);
+                }
+                // exit the inner loop when it reaches the breakpoint.
+                if(breakpoint == 0){
+                    break;
+                }
+                breakpoint--;
+            }
+
+            // exit the outermost loop when it reaches the breakpoint.
+            if(breakpoint == 0){
+                break;
+            }
+        }
+    }
+}
+
+
+
+```
+
+In the `crawl` method of the above code snippet, the `rootURL` is the starting point of the crawler and the `breakpoint` represents how many URLs you want your crawler to discover. 
+
+The algorithm starts by adding the root URL to the queue and the list of visited URLs. Then it reads the HTML content of the URL using the `BufferedReader` API to search for any embedded URLs. It uses the specified regex pattern to discover URLs in the HTML. If you’re new to regex, [here](https://www.section.io/engineering-education/regex-implementation-with-java/) is a tutorial I created that will get you started on implementing regex with Java.
+
+These operations will iterate until the crawler has discovered the number of URLs specified in your `breakpoint`.
+
+Here is a snippet of the main method for the application:
+
+```java
+
+
+    public static void main(String[] args) {
+        WebCrawler crawler = new WebCrawler();
+        String rootURL = "https://www.section.io/engineering-education/springboot-antmatchers/";
+        crawler.crawl(rootURL, 100);
+    }
+
+
+```
+
+The above `main` method uses the URL to one of my tutorials as the start URL and a breakpoint of 100.
+
+Below is a snapshot of the output of running the program:
+
+![web crawl sample output](/crawl-result.png)
```

Comment:
Explain this output. What does these URLs represent? Are these based on the URLs that are present within the webpage? Or like related webpage results?

Reply:
The URLs in the snapshot are some of the URLs contained in all the web pages the web crawler crawled starting from the root URL through the embedded URLs until it reached the breakpoint.

Reply:
Again - you need not explain them to me here. You must add these in the article

Reply:
I already added it to the article. You can view it in the new PR I just made where I tagged you. I was only making an additional response to you here.